plugins/experimental/carp/carp.cc

```diff
@@ -0,0 +1,713 @@
+    if (!bAddForwardedResult) {
+      TSError("Carp, error inserting '%s' header", CARP_STATUS_HEADER.c_str());
+    }
+
+    if (bIsPOSTRemap) { // if post remap, get remapped/OS Server Addr and add as header
+      const struct sockaddr* sa = TSHttpTxnServerAddrGet(txnp);
+      //        const struct sockaddr* sa = TSHttpTxnNextHopAddrGet(txnp);
+      if (sa) { // sanity check
+        struct sockaddr_storage ss;
+        memcpy(static_cast<void *> (&ss), sa, sizeof (sockaddr_storage));
+        if ((reinterpret_cast<sockaddr_in *> (&ss))->sin_port == 0) { // set port from client request URL
+          (reinterpret_cast<sockaddr_in *> (&ss))->sin_port = htons(reqUrl.getPort());
+        }
+
+        string sTemp;
+        getStringFromSockaddr(reinterpret_cast<const sockaddr *> (&ss), sTemp);
+        TSDebug(DEBUG_TAG_HOOK, "Inserting forward header with sockaddr:%s", sTemp.c_str());
+        string sSockaddr;
+        sSockaddr.reserve(32 + sizeof (sockaddr_storage) * 2);
+        for (unsigned int i = 0; i<sizeof (sockaddr_storage); i++) {
+          char val[8];
+          sprintf(val, "%02X", reinterpret_cast<const unsigned char *> (&ss)[i]);
+          sSockaddr += val;
+        }
+        sSockaddr += "/" + reqUrl.getScheme();
+        // insert carp forwarding header
+        bool bAddFwdHeaderResult = addHeader(bufp, hdr_loc, CARP_FORWARD_HEADER, sSockaddr);
+        if (!bAddFwdHeaderResult) {
+          TSError("Carp, error inserting '%s' header", CARP_FORWARD_HEADER.c_str());
+        }
+      }
+    } else { // for premap mode
+      string sScheme = reqUrl.getScheme();
+
+      if (!addHeader(bufp, hdr_loc, CARP_PREMAP_SCHEME, sScheme)) { 
+        TSError("Carp, error inserting '%s' header in premap mode", CARP_PREMAP_SCHEME.c_str());  
+      } else {
+        TSDebug(DEBUG_TAG_HOOK, "Insert client request scheme %s in premap mode", sScheme.c_str());
+      }
+    }
+    // set origin server/destination
+    //      TSHttpTxnConfigIntSet(txnp, TS_CONFIG_HTTP_SHARE_SERVER_SESSIONS, 0); // disable conn sharing due to bug when used with TSHttpTxnServerAddrSet
+    if (TSHttpTxnServerAddrSet(txnp, reinterpret_cast<const struct sockaddr *> (&node->forwardAddr)) != TS_SUCCESS) {
+      TSDebug(DEBUG_TAG_HOOK, "Error calling TSHttpTxnServerAddrSet");
+    } else {
+      // set scheme appropriately based on destination
+      TSDebug(DEBUG_TAG_HOOK, "Setting scheme to '%s'", node->getSchemeString());
+      TSUrlSchemeSet(bufp, url_loc, node->getSchemeString(), -1);
+      if (!bIsPOSTRemap) { // since we are forwarding, do not remap request and do not cache result
+        TSSkipRemappingSet(txnp, true);
+      }
+      TSHttpTxnArgSet(txnp, g_carpSelectedHostArgIndex, static_cast<void *> (node));
+      if (!bIsOwner) {
+        TSHttpTxnServerRespNoStoreSet(txnp, 1); // do not cache the response
+      } else {
+        TSHttpTxnServerRespNoStoreSet(txnp, 0); // we are replicate owner, cache response
```

Comment:
Is there a better API I should be using? I remember seeing multiple APIs for this and thought this was the right one.